nixos: add --specialisation to nixos-rebuild

Patryk27 opened the pull request:
###### Description of changes

This commit fixes a papercut in nixos-rebuild where people wanting to switch to a specialisation (or test one) were forced to manually figure out the specialisation's path and run its activation script - since now, there's a dedicated option to do just that.

This is a backwards-compatible change which doesn't affect the existing behavior, which - to be fair - might still be considered sus by some people, the painful scenario here being:

- you boot into specialisation `foo`,
- you run `nixos-rebuild switch`,
- whoops, you're no longer at specialisation `foo`, but you're rather brought back to the base system.

(it's especially painful for cases where specialisation is used to load extra drivers, e.g. Nvidia, since then launching `nixos-rebuild switch`, while forgetting that you're inside a specialisation, can cause some parts of your system to get accidentally unloaded.)

I've tried to mitigate that by improving specialisations so that they create a dedicated file somewhere in `/run/current-system` containing the specialisation's name (which `nixos-rebuild` could then use as the default value for `--specialisation`), but I haven't been able to come up with anything working (plus it would be a breaking change then).

Closes https://github.com/NixOS/nixpkgs/issues/174065

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Patryk27 committed:
nixos: add --specialisation to nixos-rebuild

This commit fixes a papercut in nixos-rebuild where people wanting to
switch to a specialisation (or test one) were forced to manually figure
out the specialisation's path and run its activation script - since now,
there's a dedicated option to do just that.

This is a backwards-compatible change which doesn't affect the existing
behavior, which - to be fair - might still be considered sus by some
people, the painful scenario here being:

- you boot into specialisation `foo`,
- you run `nixos-rebuild switch`,
- whoops, you're no longer at specialisation `foo`, but you're rather
  brought back to the base system.

(it's especially painful for cases where specialisation is used to load
extra drivers, e.g. Nvidia, since then launching `nixos-rebuild switch`,
while forgetting that you're inside a specialisation, can cause some
parts of your system to get accidentally unloaded.)

I've tried to mitigate that by improving specialisations so that they
create a dedicated file somewhere in `/run/current-system` containing
the specialisation's name (which `nixos-rebuild` could then use as the
default value for `--specialisation`), but I haven't been able to come
up with anything working (plus it would be a breaking change then).

Closes https://github.com/NixOS/nixpkgs/issues/174065

RaitoBezarius commented:
Could you provide with the failed NixOS test so we can maybe help to fix it together?

ncfavier commented on pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:
```diff
@@ -107,6 +108,14 @@ while [ "$#" -gt 0 ]; do
         fi
         shift 1
         ;;
+      --specialisation-name|-c)
```
`--specialisation` seems unambiguous

Patryk27 replied:
I like `--specialisation` more as well, but I wanted for it to be symmetric with `--profile-name` 👀

ncfavier replied:
No need IMO. `--profile` could possibly a path, but `--specialisation` is unambiguous

Patryk27 commented:
> Could you provide with the failed NixOS test so we can maybe help to fix it together?

I've gotten stuck with the basic idea of the test, since it looks like I can't just run `nixos-rebuild switch` there (because VM's `/nix/store` is read-only + it doesn't have any channels configured); I've taken a look at at `nixos/tests/installer.nix` and `nixos/tests/os-prober.nix`, but I've failed to understand the magic that makes them work 😅 

So I guess something very basic could be:
```nix
import ./make-test-python.nix {
  name = "nixos-rebuild";

  nodes =  {
    machine = { ... }: {
      # 
    };
  };

  testScript = ''
    machine.wait_for_unit("default.target")

    # TODO create /etc/configuration.nix (+ hardward-configuration.nix?) with
    #      multiple specializations; make following work:
    machine.succeed("nixos-rebuild switch --specialisation-name foo")
  '';
}
```

RaitoBezarius commented:
> > Could you provide with the failed NixOS test so we can maybe help to fix it together?
> 
> I've gotten stuck with the basic idea of the test, since it looks like I can't just run `nixos-rebuild switch` there (because VM's `/nix/store` is read-only + it doesn't have any channels configured); I've taken a look at at `nixos/tests/installer.nix` and `nixos/tests/os-prober.nix`, but I've failed to understand the magic that makes them work sweat_smile
> 
> So I guess something very basic could be:
> 
> ```nix
> import ./make-test-python.nix {
>   name = "nixos-rebuild";
> 
>   nodes =  {
>     machine = { ... }: {
>       # 
>     };
>   };
> 
>   testScript = ''
>     machine.wait_for_unit("default.target")
> 
>     # TODO create /etc/configuration.nix (+ hardward-configuration.nix?) with
>     #      multiple specializations; make following work:
>     machine.succeed("nixos-rebuild switch --specialisation-name foo")
>   '';
> }
> ```

If I provide you with a working baseline for the test, can you write the Python script? You can check `systemd-boot.nix` test which is interesting for your usecases.

Patryk27 commented:
I know how those tests work in general (and I've written a few before), I'm just having a problem with this particular setup - so if you manage to provide some foundation, I should be able to work on that 🙂

ncfavier commented:
> but I've failed to understand the magic that makes them work

I *think* the magic you're looking for is `installation-device.nix`? That should at least take care of channels.

https://github.com/NixOS/nixpkgs/blob/ead8d4e9a04a6c821ff0d4b893745ffa6cb74bbe/nixos/tests/installer.nix#L286-L287

Patryk27 commented:
Okie, thanks, I'll look into it 🙂

Patryk27 commented:
Okie, I've changed `--specialisation-name` into just `--specialisation` and added a test 🎉 

The test still needs a minor clean-up (I've just copy-pasted `system.extraDependencies` from `os-prober.nix` and I feel like they could get shortened), but I think it's mostly ready.

Patryk27 commented:
kk, test minimized 🙂

RaitoBezarius commented:
@ofborg test nixos-rebuild-specialisations

RaitoBezarius commented:
The aarch64 failure seems unrelated to this PR.

RaitoBezarius reviewed (approved):
Great job!

ncfavier commented on nixos/tests/all-tests.nix:
```diff
@@ -419,7 +419,6 @@ in {
   netdata = handleTest ./netdata.nix {};
   networking.networkd = handleTest ./networking.nix { networkd = true; };
   networking.scripted = handleTest ./networking.nix { networkd = false; };
-  specialisation = handleTest ./specialisation.nix {};
```
Does the new test test everything the old test tested?

Patryk27 replied:
Even more, since the old test had hard-coded paths to specializations (e.g. `/run/current-system/specialisation/inheritconf/bin/switch-to-configuration test`); but in general yeah, we don't lose anything in here.

RaitoBezarius commented:
@Patryk27 I think this PR is good to go, there's some style changes to be done on it (trailing whitespace), can I let you take care of them?

Patryk27 commented:
Oh my, fixed! 😇

donovanglover commented:
Would it be possible to make `--specialisation` work for `build-vm` as well? I assume some logic needs to be added to `nixBuild`/`nixFlakeBuild`?

Patryk27 commented:
On the top of my head, I think it should be possible; I myself have no time to prepare such change now, though :cry:

nh2 commented:
When passing a specialisation that doesn't exist, there is a lack of error message:

* https://github.com/NixOS/nixpkgs/issues/412749
